src/hal/src/pso/descriptor.rs

```diff
@@ -33,38 +33,67 @@ pub type DescriptorBinding = u32;
 ///
 pub type DescriptorArrayIndex = usize;
 
-/// DOC TODO: Grasping and remembering the differences between these
-///       types is a tough task. We might be able to come up with better names?
-///       Or even use tuples to describe functionality instead of coming up with fancy names.
-#[repr(C)]
-#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+/// Specific type of a buffer.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum BufferDescriptorType {
+    /// Storage buffers allow load, store, and atomic operations.
+    Storage {
+        /// If true, store operations are not permitted on this buffer.
+        read_only: bool,
+    },
+    /// Uniform buffers allow only load operations.
+    Uniform,
+}
+
+/// Format of a buffer.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum BufferDescriptorFormat {
+    /// The buffer is interpreted as a structure defined in a shader.
+    Structured {
+        /// If true, the buffer is accessed by an additional offset specified by
+        /// `DescriptorSetWrite::array_offset`.
```

Comment:
the comment is incorrect, the offset this refers to is actually coming from the last parameter of [bind_graphics_descriptor_sets](https://docs.rs/gfx-hal/0.4.1/gfx_hal/command/trait.CommandBuffer.html#tymethod.bind_graphics_descriptor_sets) (and the corresponding compute one)